torchtext/experimental/datasets/text_classification.py

```diff
@@ -87,32 +52,47 @@ def get_vocab(self):
         return self.vocab
 
 
-def _setup_datasets(dataset_name, root='.data', ngrams=1, vocab=None,
-                    tokenizer=None, data_select=('train', 'test')):
-    text_transform = []
+def _setup_datasets(
```

Comment:
Let's consider if we want to keep some functions (or transforms) separated from `_setup_datasets` func. For example, `ngrams_func` defines a "contract" to convert a list of tokens to ngrams. `vocab_func` defines a "contract" from tokens to ids. Although those functions are still quite preliminary, those "contracts" are reusable in the future. When we re-write the datasets (for example, language modeling datasets), we need some similar "contracts" (like tokenizer, vocab). Then, we can unify those "contracts" and add them as `data.transforms`. This is one of the reasons that I want to keep them away from `_setup_datasets`.
Another thing is the layout of `_setup_datasets`. Within the original `_setup_datasets` func, it just combines the "transforms" together as a pipeline. IMO, it's easier for users to follow if they want to create their own custom `_setup_datasets`.
Overall, I think those building blocks separated from `_setup_datasets` func may add some flexibility.

Reply:
Sure, if we want to create a collection of functions that behave according to some contract and have those as user-facing, I definitely think there's value to it. This is similar to get_tokenizer or ngrams_iterator within utils. We just might want to pick a better path. Did you mean that you'd rather have these functions next to the datasets?

Reply:
For now, since we only experiment those utils here, we could keep those utils next to the datasets. Once we have more utils for language modeling and question-answer datasets, we could start unifying them. Or we could create some experimental utils in the experimental folder, which will evolve over time. Either way works for me, IMO.

Reply:
I'd say we build out the folder structure we want to see at the module level within experimental so it'll be an easier switch / merge later on.

Reply:
There are five generic functions introduced in the current code

vocab_func - returns a function that calls ```__getitem__``` on each entry of a given list using a particular vocab object
totensor - effectively returns torch.tensor. I think this is unnecessary.
ngrams_func - we have ```ngrams_iterator``` already
build_vocab - implements ```build_vocab_from_iterator(transforms(txt) for (_, txt) in data)```
squential_transforms - effectively [functools.reduce](https://docs.python.org/3/library/functools.html#functools.reduce) and similar to nn.Sequential for nn.Modules and torchvision also has a similar combination function.

Reply:
I agree that those five functions don't cover a lot of codes and can be replaced. However, I feel like we are telling a consistent story here with some standard components. And we keep this consistent with the other two domain libraries. And we are using the same `language` as pytext since they also use transforms. Plus, I re-use those functionals in SQuAD datasets.
https://github.com/pytorch/text/blob/edb143af512e890f3a00d3e872658f0986f3b23a/torchtext/experimental/datasets/question_answer.py#L5

Reply:
Outside of ```squential_transforms``` the other two are minimal enough that I think that it's ok to repeat them. totensor effectively returns a torch.tensor with a different default dtype. I think we should be careful here, for all these transforms really, until we have a standard and wider plans around transforms and functionals.